### PR TITLE
fix: fix snapshot consistency bug

### PR DESCRIPTION
**Pull request by JackTan25**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary
Summary about this PR
1. in `resolve_data_source()`, we use catalog to read a table, so every time we read a table, we will get the newest snapshot, it's bug. for example, 'select * from t1 join (select * from t1 as t2);', we will read t1 twice, if for the interval there is a mutation operation to change the newest snapshot, it will cause different <dyn Table> for the same table in
one sql execution. So we use ctx to read.
2. in many ut tests, we find out that they use one and the same ctx for multi sqls, so we need to update cache to read the
new snapshot
3. for fixture test, if we use execute_sql to run it, the table's default catalog and default database is 'default',not fixture.default_database,fixture.default_catalog.
- Closes #issue

- - -
This change is [Reviewable](https://reviewable.io/reviews/datafuselabs/databend/13526)

### CODE REVIEW

**Comment by JackTan25** [comment]
delete test
```sql
[2023-11-01T16:10:31Z INFO  test_replace_recluster] Ok. merge-into batch (with conflict) : [2996]
[2023-11-01T16:10:31Z INFO  test_replace_recluster] executing batch: 2997
[2023-11-01T16:10:31Z INFO  test_replace_recluster] executing merge-into batch : 2997
[2023-11-01T16:10:33Z INFO  test_replace_recluster] Ok. merge-into batch : 2997
[2023-11-01T16:10:33Z INFO  test_replace_recluster] executing merge-into (with conflict) : [2997]
[2023-11-01T16:10:33Z INFO  test_replace_recluster] Ok. merge-into batch (with conflict) : [2997]
[2023-11-01T16:10:33Z INFO  test_replace_recluster] executing batch: 2998
[2023-11-01T16:10:33Z INFO  test_replace_recluster] executing merge-into batch : 2998
[2023-11-01T16:10:35Z INFO  test_replace_recluster] Ok. merge-into batch : 2998
[2023-11-01T16:10:35Z INFO  test_replace_recluster] executing merge-into (with conflict) : [2998]
[2023-11-01T16:10:36Z INFO  test_replace_recluster] Ok. merge-into batch (with conflict) : [2998]
[2023-11-01T16:10:36Z INFO  test_replace_recluster] executing batch: 2999
[2023-11-01T16:10:36Z INFO  test_replace_recluster] executing merge-into batch : 2999
[2023-11-01T16:10:38Z INFO  test_replace_recluster] Ok. merge-into batch : 2999
[2023-11-01T16:10:38Z INFO  test_replace_recluster] executing merge-into (with conflict) : [2999]
[2023-11-01T16:10:38Z INFO  test_replace_recluster] Ok. merge-into batch (with conflict) : [2999]
[2023-11-01T16:10:38Z INFO  test_replace_recluster] ==========================
[2023-11-01T16:10:38Z INFO  test_replace_recluster] ====verify table state====
[2023-11-01T16:10:38Z INFO  test_replace_recluster] ==========================
[2023-11-01T16:10:38Z INFO  test_replace_recluster]                            
[2023-11-01T16:10:38Z INFO  test_replace_recluster]                            
[2023-11-01T16:10:38Z INFO  test_replace_recluster] number of successfully executed merge-into statements : 3000
[2023-11-01T16:10:38Z INFO  test_replace_recluster]                            
[2023-11-01T16:10:38Z INFO  test_replace_recluster]                            
[2023-11-01T16:10:38Z INFO  test_replace_recluster] CHECK: value of successfully executed merge-into statements
[2023-11-01T16:10:38Z INFO  test_replace_recluster] CHECK: value of successfully executed merge-into statements: client 3000000, server 0
[2023-11-01T16:10:38Z INFO  test_replace_recluster] CHECK: full table scanning
[2023-11-01T16:10:38Z INFO  test_replace_recluster] ===========================
[2023-11-01T16:10:38Z INFO  test_replace_recluster] ======     PASSED      ====
[2023-11-01T16:10:38Z INFO  test_replace_recluster] ===========================
[2023-11-01T16:10:38Z INFO  test_replace_recluster]                            
[2023-11-01T16:10:38Z INFO  test_replace_recluster]                            
[2023-11-01T16:10:38Z INFO  test_replace_recluster] ========METRICS============
[2023-11-01T16:10:38Z INFO  test_replace_recluster] fuse_commit_mutation_unresolvable_conflict_total : 1129.0
[2023-11-01T16:10:38Z INFO  test_replace_recluster] fuse_remote_io_read_bytes_after_merged_total : 559537429.0
[2023-11-01T16:10:38Z INFO  test_replace_recluster] fuse_remote_io_seeks_after_merged_total : 7284.0
[2023-11-01T16:10:38Z INFO  test_replace_recluster] query_merge_into_accumulate_milliseconds : [{"less_than":10.0,"count":3644.0},{"less_than":50.0,"count":3644.0},{"less_than":100.0,"count":3644.0},{"less_than":250.0,"count":3644.0},{"less_than":500.0,"count":3644.0},{"less_than":1000.0,"count":3644.0},{"less_than":2500.0,"count":3644.0},{"less_than":5000.0,"count":3644.0},{"less_than":10000.0,"count":3644.0},{"less_than":20000.0,"count":3644.0},{"less_than":30000.0,"count":3644.0},{"less_than":60000.0,"count":3644.0},{"less_than":300000.0,"count":3644.0},{"less_than":600000.0,"count":3644.0},{"less_than":1800000.0,"count":3644.0},{"less_than":null,"count":3644.0}]
[2023-11-01T16:10:38Z INFO  test_replace_recluster] query_merge_into_accumulate_milliseconds_count : 3644.0
[2023-11-01T16:10:38Z INFO  test_replace_recluster] query_merge_into_accumulate_milliseconds_sum : 17.0
[2023-11-01T16:10:38Z INFO  test_replace_recluster] query_merge_into_append_blocks_counter_total : 5646.0
[2023-11-01T16:10:38Z INFO  test_replace_recluster] query_merge_into_append_blocks_rows_counter_total : 3039550.0
[2023-11-01T16:10:38Z INFO  test_replace_recluster] query_merge_into_apply_milliseconds : [{"less_than":10.0,"count":7391.0},{"less_than":50.0,"count":8047.0},{"less_than":100.0,"count":8080.0},{"less_than":250.0,"count":8198.0},{"less_than":500.0,"count":8305.0},{"less_than":1000.0,"count":8305.0},{"less_than":2500.0,"count":8305.0},{"less_than":5000.0,"count":8305.0},{"less_than":10000.0,"count":8305.0},{"less_than":20000.0,"count":8305.0},{"less_than":30000.0,"count":8305.0},{"less_than":60000.0,"count":8305.0},{"less_than":300000.0,"count":8305.0},{"less_than":600000.0,"count":8305.0},{"less_than":1800000.0,"count":8305.0},{"less_than":null,"count":8305.0}]
[2023-11-01T16:10:38Z INFO  test_replace_recluster] query_merge_into_apply_milliseconds_count : 8305.0
[2023-11-01T16:10:38Z INFO  test_replace_recluster] query_merge_into_apply_milliseconds_sum : 89646.0
[2023-11-01T16:10:38Z INFO  test_replace_recluster] query_merge_into_deleted_blocks_counter_total : 3075.0
[2023-11-01T16:10:38Z INFO  test_replace_recluster] query_merge_into_deleted_blocks_rows_counter_total : 3009660.0
[2023-11-01T16:10:38Z INFO  test_replace_recluster] query_merge_into_matched_operation_milliseconds : [{"less_than":10.0,"count":635.0},{"less_than":50.0,"count":635.0},{"less_than":100.0,"count":635.0},{"less_than":250.0,"count":635.0},{"less_than":500.0,"count":635.0},{"less_than":1000.0,"count":635.0},{"less_than":2500.0,"count":635.0},{"less_than":5000.0,"count":635.0},{"less_than":10000.0,"count":635.0},{"less_than":20000.0,"count":635.0},{"less_than":30000.0,"count":635.0},{"less_than":60000.0,"count":635.0},{"less_than":300000.0,"count":635.0},{"less_than":600000.0,"count":635.0},{"less_than":1800000.0,"count":635.0},{"less_than":null,"count":635.0}]
[2023-11-01T16:10:38Z INFO  test_replace_recluster] query_merge_into_matched_operation_milliseconds_count : 635.0
[2023-11-01T16:10:38Z INFO  test_replace_recluster] query_merge_into_matched_operation_milliseconds_sum : 683.0
[2023-11-01T16:10:38Z INFO  test_replace_recluster] query_merge_into_matched_rows_total : 3017550.0
[2023-11-01T16:10:38Z INFO  test_replace_recluster] query_merge_into_not_matched_operation_milliseconds : [{"less_than":10.0,"count":5011.0},{"less_than":50.0,"count":5011.0},{"less_than":100.0,"count":5011.0},{"less_than":250.0,"count":5011.0},{"less_than":500.0,"count":5011.0},{"less_than":1000.0,"count":5011.0},{"less_than":2500.0,"count":5011.0},{"less_than":5000.0,"count":5011.0},{"less_than":10000.0,"count":5011.0},{"less_than":20000.0,"count":5011.0},{"less_than":30000.0,"count":5011.0},{"less_than":60000.0,"count":5011.0},{"less_than":300000.0,"count":5011.0},{"less_than":600000.0,"count":5011.0},{"less_than":1800000.0,"count":5011.0},{"less_than":null,"count":5011.0}]
[2023-11-01T16:10:38Z INFO  test_replace_recluster] query_merge_into_not_matched_operation_milliseconds_count : 5011.0
[2023-11-01T16:10:38Z INFO  test_replace_recluster] query_merge_into_not_matched_operation_milliseconds_sum : 14.0
[2023-11-01T16:10:38Z INFO  test_replace_recluster] query_merge_into_replace_blocks_counter_total : 569.0
[2023-11-01T16:10:38Z INFO  test_replace_recluster] query_merge_into_replace_blocks_rows_counter_total : 2092120.0
[2023-11-01T16:10:38Z INFO  test_replace_recluster] query_merge_into_split_milliseconds : [{"less_than":10.0,"count":8655.0},{"less_than":50.0,"count":8655.0},{"less_than":100.0,"count":8655.0},{"less_than":250.0,"count":8655.0},{"less_than":500.0,"count":8655.0},{"less_than":1000.0,"count":8655.0},{"less_than":2500.0,"count":8655.0},{"less_than":5000.0,"count":8655.0},{"less_than":10000.0,"count":8655.0},{"less_than":20000.0,"count":8655.0},{"less_than":30000.0,"count":8655.0},{"less_than":60000.0,"count":8655.0},{"less_than":300000.0,"count":8655.0},{"less_than":600000.0,"count":8655.0},{"less_than":1800000.0,"count":8655.0},{"less_than":null,"count":8655.0}]
[2023-11-01T16:10:38Z INFO  test_replace_recluster] query_merge_into_split_milliseconds_count : 8655.0
[2023-11-01T16:10:38Z INFO  test_replace_recluster] query_merge_into_split_milliseconds_sum : 0.0
[2023-11-01T16:10:38Z INFO  test_replace_recluster] query_merge_into_unmatched_rows_total : 3031000.0
[2023-11-01T16:10:38Z INFO  test_replace_recluster] ===========================
[2023-11-01T16:10:38Z INFO  test_replace_recluster]                            
[2023-11-01T16:10:38Z INFO  test_replace_recluster]                            
[2023-11-01T16:10:38Z INFO  test_replace_recluster] ======CLUSTERING INFO======
[2023-11-01T16:10:38Z INFO  test_replace_recluster] cluster_key : (to_yyyymmdd(insert_time), id)
[2023-11-01T16:10:38Z INFO  test_replace_recluster] block_count: 0
[2023-11-01T16:10:38Z INFO  test_replace_recluster] constant_block_count: 0
[2023-11-01T16:10:38Z INFO  test_replace_recluster] unclustered_block_count: 0
[2023-11-01T16:10:38Z INFO  test_replace_recluster] average_overlaps: NaN
[2023-11-01T16:10:38Z INFO  test_replace_recluster] average_depth: NaN
[2023-11-01T16:10:38Z INFO  test_replace_recluster] block_depth_histogram: {}
[2023-11-01T16:10:38Z INFO  test_replace_recluster] ===========================
```

**Comment by BohuTANG** [comment]
`make lint` on your local will make the clippy happy :\

File: src/query/service/src/test_kits/table_test_fixture.rs [hunk 1]
```diff
@@ -657,6 +659,13 @@ pub async fn do_deletion(ctx: Arc<QueryContext>, plan: DeletePlan) -> Result<()>
     Ok(())
 }
 
+pub async fn do_insert(ctx: Arc<QueryContext>, plan: Insert) -> Result<()> {
+    println!("insert");
```
Comment (BohuTANG) [inline src/query/service/src/test_kits/table_test_fixture.rs:663]:
```suggestion
```

File: src/query/sql/src/planner/binder/table.rs [hunk 2]
```diff
@@ -1010,17 +1010,19 @@ impl Binder {
...
+        _tenant: &str,
         catalog_name: &str,
         database_name: &str,
         table_name: &str,
         travel_point: &Option<NavigationPoint>,
     ) -> Result<Arc<dyn Table>> {
-        // Resolve table with catalog
-        let catalog = self.catalogs.get_catalog(tenant, catalog_name).await?;
-        let mut table_meta = catalog.get_table(tenant, database_name, table_name).await?;
+        // Resolve table with ctx
```
Comment (BohuTANG) [inline src/query/sql/src/planner/binder/table.rs:1021]: Here need add more comments to desp why we change to ctx.get_table

Comment (JackTan25) [reply src/query/sql/src/planner/binder/table.rs:1021]: ok

**Comment by BohuTANG** [comment]
Unit test failed seems related to this PR

**Comment by BohuTANG** [comment]
```
     Summary [ 203.765s] 789 tests run: 780 passed, 9 failed, 8 skipped
        FAIL [   1.530s] enterprise-query::it aggregating_index::index_refresh::test_refresh_agg_index
        FAIL [   1.304s] enterprise-query::it aggregating_index::index_refresh::test_refresh_agg_index_with_limit
        FAIL [   1.302s] enterprise-query::it aggregating_index::index_refresh::test_sync_agg_index
        FAIL [   1.105s] enterprise-query::it aggregating_index::index_scan::test_index_scan
        FAIL [   1.246s] enterprise-query::it aggregating_index::index_scan::test_index_scan_agg_args_are_expression
        FAIL [   1.152s] enterprise-query::it aggregating_index::index_scan::test_index_scan_two_agg_funcs
        FAIL [   1.117s] enterprise-query::it aggregating_index::index_scan::test_index_scan_with_count
        FAIL [   1.215s] enterprise-query::it aggregating_index::index_scan::test_projected_index_scan
        FAIL [   1.138s] enterprise-query::it storages::fuse::operations::vacuum::test_fuse_do_vacuum
```

**Comment by JackTan25** [comment]
> Summary [ 203.765s] 789 tests run: 780 passed, 9 failed, 8 skipped
>         FAIL [   1.530s] enterprise-query::it aggregating_index::index_refresh::test_refresh_agg_index
>         FAIL [   1.304s] enterprise-query::it aggregating_index::index_refresh::test_refresh_agg_index_with_limit
>         FAIL [   1.302s] enterprise-query::it aggregating_index::index_refresh::test_sync_agg_index
>         FAIL [   1.105s] enterprise-query::it aggregating_index::index_scan::test_index_scan
>         FAIL [   1.246s] enterprise-query::it aggregating_index::index_scan::test_index_scan_agg_args_are_expression
>         FAIL [   1.152s] enterprise-query::it aggregating_index::index_scan::test_index_scan_two_agg_funcs
>         FAIL [   1.117s] enterprise-query::it aggregating_index::index_scan::test_index_scan_with_count
>         FAIL [   1.215s] enterprise-query::it aggregating_index::index_scan::test_projected_index_scan
>         FAIL [   1.138s] enterprise-query::it storages::fuse::operations::vacuum::test_fuse_do_vacuum

I‘m fixing

**Comment by BohuTANG** [comment]
@JackTan25 

Hi, if the PR is ready to review please:
1. Pass at least the basic CI checks, preferably all.
2. Include a summary explaining the what the PR's did.

**Comment by JackTan25** [comment]
wait https://github.com/datafuselabs/databend/pull/13564

File: src/query/service/tests/it/storages/fuse/operations/mutation/segments_compact_mutator.rs [hunk 3]
```diff
@@ -82,6 +82,7 @@ async fn test_compact_segment_normal_case() -> Result<()> {
     append_rows(ctx.clone(), num_inserts).await?;
 
     // check count
+    ctx.evict_table_from_cache("default", "default", "t")?;
```
Comment (BohuTANG) [inline src/query/service/tests/it/storages/fuse/operations/mutation/segments_compact_mutator.rs:85]: Why need to call `evict_table_from_cache` here?

File: src/query/ee/tests/it/aggregating_index/index_scan.rs [hunk 4]
```diff
@@ -614,21 +691,29 @@ async fn test_index_scan_agg_args_are_expression_impl(format: &str) -> Result<()
...
     execute_sql(
         fixture.ctx(),
         &format!("REFRESH AGGREGATING INDEX {index_name}"),
     )
     .await?;
 
     // Query with index
+    fixture
+        .ctx()
+        .evict_table_from_cache(CATALOG, DATABASE, "t")?;
```
Comment (dantengsky) [inline src/query/ee/tests/it/aggregating_index/index_scan.rs:716]: does the correctness of testing depend on those invocations of `evict_table_from_cache`?

Comment (JackTan25) [reply src/query/ee/tests/it/aggregating_index/index_scan.rs:716]: yes, please see the summary session 2 of this pr. In real sql execution, every sql has its own ctx, but for ut test, they will execute many ctx but have the same ctx, so the later pipeline will get the old snapshot info which is from previous ones. cc @dantengsky

File: src/query/service/tests/it/storages/fuse/operations/mutation/segments_compact_mutator.rs (hunk 3 above)
Comment (JackTan25) [reply src/query/service/tests/it/storages/fuse/operations/mutation/segments_compact_mutator.rs:85]: see line 82 ,it has a append_rows, we need to update cache. You can see the summary session 2 of this pr. cc @BohuTANG

**Comment by dantengsky** [comment]
:+1: Thanks for diagnosing and fixing this bug.

but I am not sure if the test case given in `src/query/service/tests/it/sql/exec/mod.rs` can effectively detect the bug.

https://github.com/datafuselabs/databend/pull/13526/files#diff-20b999ff94b1aedd5d21d6d9047b64ac8d9ebab7bb0a60390c042247564a3535R41-R42

If not, how about mocking the Catalog and detecting if a table has been resolved multiple times using `Catalog::get_table` for the "self join" query?

for your reference:

https://github.com/datafuselabs/databend/blob/bdf21ae8f15be8fe2e6b207e6736e542aec8e73c/src/query/service/tests/it/storages/fuse/operations/commit.rs#L246-L251

**Comment by JackTan25** [comment]
> 👍 Thanks for diagnosing and fixing this bug.
> 
> but I am not sure if the test case given in `src/query/service/tests/it/sql/exec/mod.rs` can effectively detect the bug.
> 
> https://github.com/datafuselabs/databend/pull/13526/files#diff-20b999ff94b1aedd5d21d6d9047b64ac8d9ebab7bb0a60390c042247564a3535R41-R42
> 
> If not, how about mocking the Catalog and detecting if a table has been resolved multiple times using `Catalog::get_table` for the "self join" query?
> 
> for your reference:
> 
> https://github.com/datafuselabs/databend/blob/bdf21ae8f15be8fe2e6b207e6736e542aec8e73c/src/query/service/tests/it/storages/fuse/operations/commit.rs#L246-L251

oh, sorry, I think we just need to test current new logic (`use ctx`). I forget to test the old logic bug. I think we need to preserve the test in mod.rs and add another new test.

**Comment by dantengsky** [comment]
> ... But if we test catalog, maybe we will get a flakey test for future ci ut test, can we accept this? cc @dantengsky

If the mocked objects are kept within the unit test, they should not cause flakiness, assuming I understand your concerns correctly.

File: src/query/ee/tests/it/aggregating_index/index_scan.rs (hunk 4 above)
Comment (dantengsky) [reply src/query/ee/tests/it/aggregating_index/index_scan.rs:716]: oh, I see.

seems the related test case should be adjusted (by using new query context ?), let's keep the changes as it is, and fix them in another PR.

Comment (dantengsky) [reply src/query/ee/tests/it/aggregating_index/index_scan.rs:716]: turns out, an issue is tracking this 

https://github.com/datafuselabs/databend/issues/11887

**Review by dantengsky — APPROVED** [review]

**Pull request merged**